ussl: Fix polling of SSL sockets

damz opened the pull request:
### What does this PR solve?

There are three issues in the current implementation `ioctl(MP_STREAM_POLL)` for SSL sockets:

* mbedtls might have buffered data (which we can check with `mbedtls_ssl_check_pending`) which needs to be considered when checking for `MP_STREAM_POLL_RD`;
* During handshake (and renegotiation) the state machine might need to write in order to read (or read in order to write);
* An error raised during `read()` or `write()` needs to be stored so that we don't block on polling (e.g. we should not wait to read data on an SSL socket that is in error).

### How does this PR solve the problem?

* Store the direction the library told us it needs, and the last error when raised;
* Only consider the direction the library to us it needed during `ioctl(MP_STREAM_POLL)`;
* Implement a full unit test in unix/coverage

dpgeorge commented:
Thanks for the contribution.  This is a good, simple fix, and since #5840 has not seen any progress (due to lack of time on my side) I suggest we merge this.  Then #5840 can build on it.

damz commented:
@dpgeorge Thank you for pointing me in the right direction. After this I can send a separate PR implementing the write-for-read / read-for-write polling logic.

After this, the reminder of #5840 is the error logic change, but unfortunately I cannot say I understand why it is necessary.

dpgeorge commented:
> After this I can send a separate PR implementing the write-for-read / read-for-write polling logic.

If it will only copy the work done in #5840, then it would be better to merge #5840 itself.  The reason I didn't merge that PR yet is because I don't fully understand that logic.  Maybe you can implement it in a simpler way?

> the error logic change, but unfortunately I cannot say I understand why it is necessary

Let's leave that out for now.

damz commented:
I think you really just need this: https://github.com/damz/micropython/commit/1166b8661b0ff26087e58f9b6ab369b8cda7b96f

The library tells you if it needs to read next or to write next. That is the only direction you are interested in for the next poll operation. Which operation you reenter after that doesn't matter, as both read and write paths reenter the handshake/renegotiation process.

dpgeorge commented:
> I think you really just need this

That is definitely simpler than #5840.  Did you have any test cases that show that your patch does indeed work?

dpgeorge commented on extmod/modussl_mbedtls.c:
```diff
@@ -366,6 +366,14 @@ STATIC mp_uint_t socket_ioctl(mp_obj_t o_in, mp_uint_t request, uintptr_t arg, i
         mbedtls_entropy_free(&self->entropy);
+    } else if (request == MP_STREAM_POLL) {
+        /* For POLL_RD, first check if ssl layer has bytes available... */
+        if (arg & MP_STREAM_POLL_RD) {
```
What about the case when `arg` has both `MP_STREAM_POLL_RD` and `MP_STREAM_POLL_WR` set (and `mbedtls_ssl_get_bytes_avail` returns >0)?  Is it fair enough to shortcut the poll test and just return `MP_STREAM_POLL_RD` immediately, like is done in this PR?  Or should it also poll the underlying stream for writability, to potentially return both RD and WR flags set?

Some badly written code might poll for both RD and WR, but only act on WR.  Such code would not work correctly with this commit here because WR will potentially never be set.

OTOH, most code (eg uasyncio) polls either RD or WR exclusively, or if it does poll both RD and WR then it acts on whichever of those is set.  And shortcutting the check is obviously faster than checking the underlying stream as well.

dpgeorge replied:
I'm sure someone will write such code as some point, so we should make it work otherwise it'll lead to subtle bugs.

Since the case of polling only for RD or WR but not for both is common, I suggest doing the shorcut only if `arg == MP_STREAM_POLL_RD`.  Otherwise fall through to call the underlying socket, and "or" the result of the underlying socket with `MP_STREAM_POLL_RD` in the case mbedtls has buffered bytes ready for reading.

damz replied:
Thank you, that makes sense. I updated the PR and incorporated the poll mask changes into it.

damz replied:
Testing this is tricky, as the `unix` port only supports polling on file descriptors and just completely bypasses this code... I am trying to put together a "multi" test setup.

dpgeorge replied:
If you can write a network multitest that would be great.  Otherwise just a normal test in `tests/net_inet` or `tests/net_hosted` would be fine.

damz replied:
@dpgeorge Would it be possible / acceptable to expose a Python interface to be able to call `->ioctl` on an object implementing the stream protocol? That would allow writing a unit test for this that works on all ports.

dpgeorge replied:
You should be able to call it using `select.select` and/or `select.poll`.  You can also create a Python object inheriting from `io.IOBase` and implementing `def ioctl(self, request, arg)` which can then be passed to `ssl.wrap_socket()`.

What exactly do you want to do?

damz replied:
You can do that, but it is unfortunately not useful on the `unix` port because it doesn't call `ioctl(MP_STREAM_POLL)` at all (it only calls `ioctl(MP_STREAM_GET_FILENO)` and polls the underlying file descriptor directly).

dpgeorge commented on extmod/modussl_axtls.c:
```diff
@@ -277,6 +277,19 @@ STATIC mp_uint_t ussl_socket_ioctl(mp_obj_t o_in, mp_uint_t request, uintptr_t a
+        }
+
+        // ...otherwise fall through to pass request to underlying socket
+        mp_uint_t ret = mp_get_stream(self->sock)->ioctl(self->sock, request, arg, errcode);
```
To reduce code size, it would be better to have just one call to `->ioctl()`, ie combine this call with the one at the end of the function.

dpgeorge commented on extmod/modussl_mbedtls.c:
```diff
@@ -366,8 +375,40 @@ STATIC mp_uint_t socket_ioctl(mp_obj_t o_in, mp_uint_t request, uintptr_t arg, i
+        }
+
+        // ...otherwise fall through to pass request to underlying socket
+        mp_uint_t ret = mp_get_stream(self->sock)->ioctl(self->sock, request, arg, errcode);
```
Can you make it so there's only one of these calls to `->ioctl()` in this function?  For example, at the very end of this function do `return ret | mp_get_stream(...)->ioctl(...);`, and then make sure `ret` is correct for all paths of logic flow.

damz replied:
I don't see how we could do that, given that `ret` would have to be conditional on the result of the `->ioctl()` call.

dpgeorge replied:
How about something like this:
```c
    mp_uint_t ret = 0;
    if (request == MP_STREAM_CLOSE) {
        ...
    } else if (request == MP_STREAM_POLL) {
        <existing logic of this if branch up to this ioctl call>
        if (has_pending) {
            ret |= MP_STREAM_POLL_RD;
        }
    }
    ret |= mp_get_stream(self->sock)->ioctl(...);
    if (request == MP_STREAM_POLL) {
        if (ret & self->poll_mask) {
            ret |= saved_arg;
        }
    }
    return ret;
}
```
?

Note that this is purely for code size reduction, so only worth doing if it reduces compiled firmware size.

damz replied:
Thank you, I adopted this in the last version.

damz committed:
extmod/modussl_mbedtls: Fix support for ioctl(MP_STREAM_POLL).

During the initial handshake or subsequent renegotiation, the protocol
might need to read in order to write (or conversely to write in order
to read). It might be blocked from doing so by the state of the
underlying socket (i.e. there is no data to read, or there is no space
to write).

The library indicates this condition by returning one of the errors
`MBEDTLS_ERR_SSL_WANT_READ` or `MBEDTLS_ERR_SSL_WANT_WRITE`. When that
happens, we need to enforce that the next poll operation only considers
the direction that the library indicated.

In addition, mbedtls does its own read buffering that we need to take
into account while polling, and we need to save the last error between
read()/write() and ioctl().

damz commented:
@dpgeorge I updated the code as you requested, and based on further experience with running this code over the last few weeks at Ribbit Network.

There is now a full unit test that runs in unix/coverage.

damz commented:
As recommended by @Carglglz I reused the certificate/key pair already used in `ssl_data.py`.

dpgeorge commented:
Thank you very much for the work on this, it's a nice clean change with a good test.
